apps/src/dance/lab2/ProgramExecutor.ts

```diff
@@ -103,28 +103,36 @@ export default class ProgramExecutor {
    * Preview the program. Compiles student code and calls on the native API to draw a frame.
    */
   async preview() {
-    this.nativeAPI.setForegroundEffectsInPreviewMode(true);
     this.reset();
-    this.hooks = await this.preloadSpritesAndCompileCode(
-      this.getCode(),
-      'runUserSetup'
-    );
+
+    await this.preloadSprites(this.getCode());
+    this.hooks = this.compileCode(this.getCode(), 'runUserSetup');
 
     if (!this.hooks.runUserSetup) {
       Lab2MetricsReporter.logWarning('Missing required hook in compiled code');
       return;
     }
 
-    this.hooks.runUserSetup();
+    const previewDraw = () => {
+      this.nativeAPI.setEffectsInPreviewMode(true);
+
+      // the user setup hook initializes effects,
+      // needs to happen in preview mode for some effects (eg, tacos)
+      if (!this.hooks.runUserSetup) {
+        return;
+      }
```

Comment:
nit: this should be covered by the check above right? should we just move the logWarning here and remove the previous check?

Reply:
yeah, TS demanded that I cover the case that it doesn't exist. I think we do want to keep the check outside of this callback so that we don't end up in a weird preview mode = true state if the hook wasn't available for some reason?